[Draft] [WIP] Support obtaining subsets from cloud-hosted FITS files

barentsen opened the pull request:
**This is a draft PR posted in the `fornax-navo/astropy` repository for preliminary review. It is my intention to close this PR shortly in favor of opening a proper PR in the `astropy/astropy` repo.**

## Summary

This PR would enable users to efficiently extract subsets of data from FITS files stored on a web server or in the cloud.  Specifically, this PR proposes adding the `use_fsspec` and `fsspec_kwargs` parameters to `astropy.io.fits.open` to enable remote FITS to be accessed with random access reads using the [fsspec package](https://filesystem-spec.readthedocs.io/).

This would enable the following usage pattern:

```python
from astropy.io import fits

# URI of a 213 MB FITS file hosted in Amazon AWS S3 cloud storage
uri = "s3://stpubdata/hst/public/j8pu/j8pu0y010/j8pu0y010_drc.fits"

# Download the primary header
with fits.open(uri, use_fsspec=True) as hdul:

    # Download a single header
    header = hdul[1].header

    # Download a single image
    mydata = hdul[1].data

    # Download a small cutout
    myslice = hdul[2].section[10:12, 20:22]
```

**A key feature is that the example above does not download the entire 213 MB file. Instead, only the necessary chunks of the FITS file are transferred on demand.**

## How does this work?

The efficient behavior is achieved by using the well-established [fsspec package](https://filesystem-spec.readthedocs.io) to open remote files.  Fsspec provides seamless access to cloud data by providing a `file`-like interface to a range of remote file systems.  In the case of files hosted on web servers or in cloud storage, fsspec translates random access `file.read()` operations into buffered [HTTP Range Requests](https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests).

Once a remote FITS file has been opened with fsspec, `astropy.io.fits` can use it in the same way as a local file. As a result, we can leverage two existing "lazy data loading" features which have been part of Astropy for many years:
1. The `lazy_load_hdus=True` parameter takes care of reading HDU header and data sections on demand rather than loading all HDUs at once.
2. The `ImageHDU.section` property enables a subset of a data array to be read into memory without downloading the entire image or cube.

Fsspec is already a dependency of major packages such as `dask` and `pandas`, so we can reasonably expect it to be maintained long term.  For example, pandas uses fsspec to enable users to open data from S3 using `pandas.read_csv("s3://...")`.

## Which changes does this PR make?

The changes required to Astropy are fairly modest.  In summary, this PR:
* Adds `fsspec` as an optional dependency, alongside the fsspec-affiliated packages `s3fs` and `gcsfs` (for Amazon and Google Cloud storage, respectively).
* Adds the  `use_fsspec` parameter, which triggers file paths to be opened with `fsspec.open()` rather than `open()`.
* Adds the `fsspec_kwargs` parameter, which enables buffering options and cloud credentials to be configured.
* Adds a new docs chapter titled "Obtaining subsets from cloud-hosted FITS files" (`docs/io/fits/usage/cloud.rst`).
* Adds unit tests (`io/fits/tests/test_fsspec.py`).
* Makes minor edits to the docs and the implementation of `.section`.

## How can we verify it really works?

I wrote a lightweight tool called [fsspec-monitor](https://github.com/barentsen/fsspec-monitor) which enables the exact network traffic to be explored when opening a FITS file with astropy+fsspec.  For example, we can use this tool to monitor what happens when we request a 10-by-20 pixel cutout from a 213 MB FITS file as follows:
```python
from fsspecmonitor import FsspecMonitor  # requires `barentsen/fsspec-monitor`
from astropy.io import fits

# URL of a 213 MB Hubble image
url = "https://mast.stsci.edu/api/v0.1/Download/file/?uri=mast:HST/product/j8pu0y010_drc.fits"

with FsspecMonitor() as monitor:
    with fits.open(url, use_fsspec=True, fsspec_kwargs={"block_size": 500_000}) as hdul:
        cutout = hdul[2].section[10:20, 30:40]
    monitor.summary()
```

The above example triggers the following diagnostic log to be written stdout:
```
Reading https://mast.stsci.edu/api/v0.1/Download/file/?uri=mast:HST/product/j8pu0y010_drc.fits (213.82 MB)
Fetch bytes 0-500080 (0.91 MB/s)
Fetch bytes 74759040-75261920 (2.71 MB/s)
Summary: fetched 1002960 bytes (0.96 MB) in 0.70 s (1.36 MB/s) using 2 requests.
```

These diagnostics confirm that it took just 2 HTTP requests and less than 1 MB of data to extract a cutout from a 213 MB FITS file. 🎉 

(Disclosure: I optimized this example by using `fsspec_kwargs` to set the minimum `block_size` to 500KB.  The block size refers to the fact that fsspec uses a buffered readahead strategy to trade-off the number of requests on one hand and the amount of data transferred on the other.  The minimum block size is 5MB by default and would have caused 10MB of data to be transferred in this example.)

barentsen committed:
Initial prototype

barentsen committed:
Early optimizations

barentsen committed:
Enable seamless opening of S3 URIs

barentsen committed:
Support single slice

barentsen committed:
Add initial unit test

barentsen committed:
Add additional tests

barentsen committed:
Add unit tests and bugfixes

barentsen committed:
Add comments and additional tests

barentsen committed:
Add docstrings

barentsen committed:
Add test case for compressed images

barentsen committed:
Improve docstrings

barentsen committed:
Add S3 unit test

barentsen committed:
Make s3fs an optional dependency

barentsen committed:
Improve fsspec usage

barentsen committed:
Introduce the optional parameter use_fsspec

barentsen committed:
Add preliminary change log entry

barentsen committed:
Implement simplified repr

barentsen committed:
Add test for local file opening with fsspec

barentsen committed:
Improve docstrings

barentsen committed:
SubsetRetriever => LazyLoadArray

barentsen committed:
Add test for Cutout2D

barentsen committed:
Improve docstrings

bsipocz commented:
> This PR is a work in progress and will likely require significant discussion and review. For example, I would love to understand whether the community believes this functionality belongs in AstroPy core, or whether it belongs in a "cloud optimized fits reader" affiliate package.

💯  that it belongs to core astropy, there have been readers with much fewer potential use cases, and this has the potential to be useful for any of the data providers moving towards data lakes and the cloud.

bsipocz commented:
> Is fsspec sufficiently mature to be added as an (optional) AstroPy dependency?

definitely, as you said big foundational packages already rely on it, I don't see any red flags of why it couldn't be added as an optional dependency.

bsipocz commented on astropy/io/fits/hdu/image.py:
```diff
@@ -214,6 +215,149 @@ def header(self, header):
+        dimension.  For example, an image of shape (100, 200) is defined to be
+        composed of 100 chunks.  As a result, slicing operations which limit
+        the data required along the first dimension will be more efficient.
+        For example, `.subset[10:12, :]` would trigger two chunks to be
```
these links will fail the CI, I would suggest running the tests and build the docs on this locally before you open PR.

barentsen replied:
Agreed!  Will do.

barentsen replied:
This should be resolved now -- I attempted to add extensive docstrings and narrative docs!

bsipocz commented on astropy/io/fits/hdu/image.py:
```diff
@@ -214,6 +215,149 @@ def header(self, header):
+            >>> f = fits.open(uri, fsspec_kwargs={'default_cache_type': "block", 'default_block_size': 5_000_000})
+        """
+
+        class LazyLoadArray():
```
I feel it may make sense to keep the reference to HDUs, maybe even image HDUs in the name of this class as it's not generic for any arrays

barentsen replied:
Agreed!

barentsen replied:
(I also intend to move this lazy array class somewhere else, TBD, as it's defined inline now!)

barentsen replied:
I ended up being able to re-use the `ImageHDU.section` property rather than implementing a new `ImageHDU.subset` attribute.  As a result, this class is no longer needed and has been removed!

bsipocz commented on astropy/io/fits/hdu/image.py:
```diff
@@ -214,6 +215,149 @@ def header(self, header):
+        class LazyLoadArray():
+
+            def __init__(self, imagehdu, shape, dtype, offset, *args, **kwargs):
+                self.imagehdu = imagehdu
```
you will need docstrings

barentsen replied:
Agreed!

bsipocz commented on astropy/io/fits/hdu/hdulist.py:
```diff
@@ -105,6 +105,22 @@ def fitsopen(name, mode='readonly', memmap=None, save_backup=False,
+        In addition, Amazon S3 files also require the ``s3fs`` optional
+        dependency, and Google Cloud files require ``gcsfs``.
+
+        .. versionadded:: 5.1
```
we're already in the 5.2 dev cycle

```suggestion
        .. versionadded:: 5.2
```

barentsen replied:
Merged the suggestion! :+1:

bsipocz reviewed (commented):
Some nitpicking comments. Two main points: this will need docs, both docstrings and narrative. And you may want to run tests and docs build locally before pushing it as a PR to cut out a few rounds of necessary fixes.

barentsen committed:
Update astropy/io/fits/hdu/hdulist.py

Co-authored-by: Brigitta Sipőcz <b.sipocz@gmail.com>

barentsen committed:
Rename LazyLoadArray => LazyImageData

barentsen committed:
Fix codestyle

barentsen committed:
Add docs placeholders

barentsen committed:
Fix versionadded

barentsen committed:
First stab at narrative docs

barentsen committed:
Update docstring of `section`

barentsen committed:
Improve fsspec_kwargs defaults

barentsen committed:
Update Section object

barentsen committed:
Switch .subset to .section

barentsen committed:
Retire .subset in favor of .section

barentsen committed:
Oops, the example gallery is deprecated

barentsen committed:
Minor tweaks

barentsen committed:
Improve defaults

barentsen committed:
Improve docstrings

barentsen committed:
Update tests

barentsen committed:
Update tests and docstrings

barentsen committed:
Narrative docs draft

barentsen committed:
Doc improvements

m-stclair commented:
Very fun! I definitely think it is ready to submit to astropy. I just 
have a few notes:

1. I really like the locally-defined `LazyLoadArray` class. I have one 
   nitpicky comment about it from a fault-tolerance perspective. I think you 
   should do one of the following:
   1. bind the shape, dtype, and offset of the parent ImageHDU in a closure 
   rather than using `LazyLoadArray.__init__`
   2. prevent the `LazyLoadArray` object from ever being returned to the enclosing scope of the
   `ImageHDU`.
   3. ignore assignments to these attributes, or -- I think less usefully -- ensure that
   they work 'properly', i.e. casting to the new dtype.

    At present, it is easy to accidentally do something like:
   ```
   subset = hdu.subset  # hdu's real dtype is 8-bit integer
   subset.dtype = np.float64
   subset[10:20]
   ```
   And receive a bizarre `TypeError`.

2. From an API design perspective, it is somewhat confusing for ImageHDU to 
have both a .subset property and an (apparently obsolete) .section property. 
If .section is in fact mostly obsolete, .subset might want to simply replace 
or merge with .section, with added hooks as necessary for backwards compatibility.
But this is really a question for the astropy maintainers.
3. Do you have any implementation ideas for subsetting at a more granular 
level than grabbing full "lines"? That's definitely the
most efficient thing in terms of total number of requests, of course, but 
for "wide" arrays, or arrays > 2D, it will often end up grabbing most of the 
data volume to get a very small subset. This is partly just a limitation of 
any classic raster format, of course, and is the problem things like FITS 
tile compression and Zarr were made to solve...but in cases like a 1000x10 
slice from a 1000x40000 array, it would probably be worth making some extra 
non-contiguous requests to avoid grabbing the extra bytes.
4. A related note: Decreasing `fsspec`'s default blocksize seems strictly 
worse in cases where you are not saving many bytes at the head or tail of a 
chunk: more requests for little benefit. I might recommend wrapping this in 
some way to automatically modify `fsspec`'s blocksize in those cases.

barentsen committed:
Improve docstrings

m-stclair commented:
Glancing at this again, I think you had maybe already addressed my point 2 in a branch other than the one I was looking at -- sorry for any confusion. I like what you did with section.

barentsen committed:
Improve narrative docs

barentsen committed:
Add use_fsspec as a configuration item

barentsen committed:
Improve narrative docs

barentsen commented:
@m-stclair Thank you so much for the interesting and helpful review!  It contributes to several insights and improvements explained below.

> From an API design perspective, it is somewhat confusing for ImageHDU to have both a .subset property and an (apparently obsolete) .section property. 

Thank you for drawing attention to this.  It wasn't until *after* I had implemented the `ImageHDU.subset` attribute that I started to appreciate that the existing `ImageHDU.section` property was doing the exact same thing.

(Indeed the Astropy docs had let me astray by mentioning that `.section` had been made obsolete by memory mapping -- which is true only for local files.  Moreover, I didn't thoroughly understand the logic implemented in `.section` until after I had gone through the mental process of re-implementing it.)

**I believe this is good news: it means we can re-use the well-established `.section` property!**

As you have noticed in the meantime, I removed my implementation of `.subset` from this PR in favor of testing and documenting the use of `fsspec` with the existing `.section` property. Put differently, this PR is now more focused on modifying the Astropy docs to explain how the existing `.section` and `lazy_load_hdus` features can be leveraged in combination with `fsspec`. For example, I have now added a chapter to the `astropy.io.fits` docs section titled `Obtaining subsets from cloud-hosted FITS files`.


> Do you have any implementation ideas for subsetting at a more granular level than grabbing full "lines"? 

It is conceivable to add this behavior to Astropy -- we'd "only" have to modify the internal implementation of `.section`.  This is not an easy task of course.  Accounting for multi-dimensional data strides in a generic and elegant way seems like a daunting challenge, but I expect it can be done without changing the API.

Fortunately, the existing "grabbing full lines" behavior already provides a theoretical 100x data transfer reduction for the common use case of extracting a 20x20 cutout from a 2000x2000 image.  As a result, I wouldn't want this PR to get stuck on trying to implement the more granular data grabbing, though I 💯% think we should attempt prototyping it in the coming weeks.


> A related note: Decreasing fsspec's default blocksize seems strictly worse in cases where you are not saving many bytes at the head or tail of a chunk: more requests for little benefit. I might recommend wrapping this in some way to automatically modify fsspec's blocksize in those cases.

I believe some of the inefficient behavior you observed may have been caused by the fact that a previous version of this PR enforced `fsspec` to use `default_cache_type="block"` rather than `default_cache_type="bytes"` (which is the default).  The "block" strategy strictly only makes `blocksize` requests at a time, whereas the "bytes" mode implements a smarter read-ahead strategy which allows larger contiguous requests.

The more I experiment with this code, and the more I learn about your experiments, the more I appreciate how the exact behavior and performance of the GET requests appears to be a complicated interplay between:
* the nature of the data product and cutout desired;
* the throughput and latency of the network;
* the way Astropy issues `file.read()` calls;
* the `fsspec` blocksize configured;
* the `fsspec` read and caching strategy (`fsspec` offers a handful different read/caching options specified via `cache_type`).

Because this is such a complicated interplay, my preferred way forward right now is to adopt the `fsspec` default settings, document the importance of the `block_size` and `cache_type` parameters, and discuss the trade-offs involved in an accessible way as part of the Astropy docs chapter that's being drafted as part of this PR.  This would offer a resource that works very well for the "extracting a 20x20 cutout from a 2000x2000 image" use case, whilst also providing a framework to experiment with more challenging use cases. 

Does any of this make sense?  Happy to discuss over voice!

barentsen committed:
Docs improvements

barentsen commented:
@falkben Good news: I believe this PR branch is working very well for TESS cube cutouts.

Using the [fsspec-monitor](https://github.com/barentsen/fsspec-monitor) tool, I was able to verify that extracting an n-by-m cutout from a TESS cube requires the expected number of HTTP GET requests (`n+1`) and the expected volume of data transferred (`(n+1)*block_size`).

By default, `fsspec/s3fs` uses a minimum `block_size` of 5 MB, so I find that a 4x6 cutout yields 5 requests and 25 MB of data transferred.  Example:

<img width="795" alt="Screen Shot 2022-05-08 at 6 34 25 PM" src="https://user-images.githubusercontent.com/817669/167326458-cb250be4-1a7f-4371-9d45-bebaddb3c17c.png">

 
The default 5 MB `block_size` is a poor choice in this example, because we only need to fetch about ~170 KB for every row in the cutout.  Reducing the minimum `block_size` to a value less than 170 KB yields superior performance.  Example:

<img width="795" alt="Screen Shot 2022-05-08 at 6 37 25 PM" src="https://user-images.githubusercontent.com/817669/167326492-548a7890-5eca-4c5d-8e91-46b71ceab00a.png">

 
It is not obvious that Astropy can optimize the `block_size` automatically because the ideal value is dependent on the exact slice(s) requested by the user.  Fortunately, in the specific case of TESScut, we know that n-by-m cutout should work well with

`block_size <= m * hdul[1].section.shape[2] * hdul[1].section.shape[3] * 4 bytes`

...so we have an easy way to optimize TESScut! 🎉

barentsen committed:
Documentation improvements

barentsen committed:
Documentation improvements

falkben commented on astropy/io/fits/hdu/image.py:
```diff
@@ -1198,6 +1224,8 @@ def _verify(self, option='warn'):
     def __init__(self, indx, naxis):
         if _is_int(indx):
+            if indx < 0:
+                indx = indx + naxis
```
Can we add a comment about what's going on here?

I didn't dive in to figure out how `_IndexInfo` is even used -- assume this has to do with negative indexing and needing to wrap around, but it seems strange to me that this wasn't supported before

barentsen replied:
You're absolutely right -- these two lines add support for negative indexing, which previously wasn't support by the `.section` feature.  I have now added a comment in the code to explain this.  Thank you!

falkben reviewed (commented):
Had a look through the code and this all looks great. Just a minor comment. To me, this feels ready for the PR target to be switched over.

barentsen committed:
Docs editing

barentsen committed:
Docs editing

barentsen committed:
Add comment about negative indexing

bsipocz commented on docs/io/fits/index.rst:
```diff
@@ -134,6 +134,37 @@ because by that point you are likely to run out of physical memory anyways), but
+Hubble Space Telescope image located in the Hubble's public
+Amazon S3 bucket as follows:
+
+.. doctest-remote-data::
+    # Location of a large Hubble archive image in Amazon S3 (213 MB)
+    uri = "s3://stpubdata/hst/public/j8pu/j8pu0y010/j8pu0y010_drc.fits"
+
+    # Extract a 10-by-20 pixel cutout image
+    with fits.open(uri, use_fsspec=True, fsspec_kwargs={"anon": True}) as hdul:
+        cutout = hdul[1].section[10:20, 30:50]
```
You need to prepend all examples with the appropriate `>>>` or `...`, here and below, as well as don't leave any empty lines, otherwise those are just docs lines and not picked up for testing.


```suggestion
.. doctest-remote-data::
    >>> # Location of a large Hubble archive image in Amazon S3 (213 MB)
    >>> uri = "s3://stpubdata/hst/public/j8pu/j8pu0y010/j8pu0y010_drc.fits"
    >>> 
    >>> # Extract a 10-by-20 pixel cutout image
    >>> with fits.open(uri, use_fsspec=True, fsspec_kwargs={"anon": True}) as hdul:
    ...    cutout = hdul[1].section[10:20, 30:50]
```

barentsen replied:
Ohhh, thank you.  It rendered correctly so I didn't think the `>>>`'s were necessary, but I will add them!

bsipocz replied:
yeap, rendering is a different beast, sphinx is happy to render after the `::` or directive, but pytest is more picky for proper indications for python snippets. None of these would fail as none of them would be picked up as a doctest.

barentsen replied:
Ah, thank you for explaining!  I better try running and reviewing the doctests in depth.  I will fix the missing `>>>` and `...`'s.

barentsen committed:
Update docs/io/fits/index.rst

Co-authored-by: Brigitta Sipőcz <b.sipocz@gmail.com>

barentsen committed:
Code formatting fixes flagged by @bsipocz

barentsen committed:
Fix test failure

barentsen committed:
Fix .section links

barentsen committed:
Fix docs

barentsen committed:
Docs editing

barentsen committed:
Docs editing

barentsen committed:
Docs editing

cmillion commented on docs/io/fits/usage/cloud.rst:
```diff
@@ -0,0 +1,263 @@
+.. note::
+
+    The `ImageHDU.section` feature is only available for uncompressed FITS
+    image extensions.  Attempting to use ``.section`` on a compressed
```
```suggestion
    image extensions.  This includes file-level compression like `gzip` as well as compression internal to the FITS format, like tile compression. Attempting to use ``.section`` on a compressed
```

barentsen committed:
Update docs/io/fits/usage/cloud.rst

Co-authored-by: Chase Million <1483210+cmillion@users.noreply.github.com>

barentsen commented:
Thanks for the initial feedback everyone!  I have now moved this PR to the main `astropy/astropy` repository.  I'm closing this PR to make sure we capture all discussion there.
